### PR TITLE
Add Open Category

### PR DESCRIPTION
**Pull request by vietbeu**

### CODE REVIEW

File: generation_models/open_category_pipeline.py [hunk 1]
```diff
@@ -0,0 +1,23 @@
+from PIL import Image
+from .base_model import BaseModel
+
+class OpenCategoryPipeline(BaseModel):
+    def load_model(self, supporting_pipelines, **kwargs):
+        ### Miner update code here
```
Comment (toilaluan) [inline generation_models/open_category_pipeline.py:6]: Should use `diffusers.DiffusionPipeline.from_pretrained(model_name)` as baseline. Miner should easily select model id by argument parser.

File: services/rewarding/app.py [hunk 2]
```diff
@@ -103,6 +101,15 @@ def __init__(self, model_handle: DeploymentHandle, args):
             self.allowed_ips_thread.daemon = True
             self.allowed_ips_thread.start()
 
+    async def __call__(self, reward_request: RewardRequest):
+        raise NotImplementedError("This method should be implemented by subclasses")
+
+class RewardApp(BaseRewardApp):
```
Comment (toilaluan) [inline services/rewarding/app.py:107]: May be rename to `FixedCategoryRewardApp`

File: services/rewarding/open_category_reward.py [hunk 3]
```diff
@@ -0,0 +1,275 @@
...
+from PIL import Image
+from tqdm import tqdm
+from transformers import AutoProcessor, AutoModelForCausalLM
+import torch
+import requests
+from generation_models.utils import base64_to_pil_image
+import os
+
+class DSGPromptProcessor:
+    def __init__(self, model_name="Gemma7b"):
```
Comment (toilaluan) [inline services/rewarding/open_category_reward.py:13]: set default to `Llama3_70b`

File: services/rewarding/open_category_reward.py [hunk 4]
```diff
@@ -0,0 +1,275 @@
...
+        import pyiqa
+        device = torch.device("cuda") if torch.cuda.is_available() else torch.device("cpu")
+        self.model = pyiqa.create_metric(model_name, device=device)
+    def __call__(self, image_path):
+        return self.model(image_path)
+
+class OpenCategoryReward():
+    def __init__(self):
+        self.iqa_metric = IQA(model_name="nima-vgg16-ava")
+        self.processor = DSGPromptProcessor(model_name="Llama3_70b")
```
Comment (toilaluan) [inline services/rewarding/open_category_reward.py:239]: rename to prompt_adherence_metric

File: services/rewarding/open_category_reward.py [hunk 5]
```diff
@@ -0,0 +1,275 @@
...
+        self.weights = {
+            "iqa": 0.3,
+            "prompt_adherence": 0.7
+        }
+    @staticmethod
+    def normalize_score(scores, min_val, max_val):
+        """Normalize the score to a 0-1 range."""
+        normalized_scores = [(score - min_val) / (max_val - min_val) for score in scores]
```
Comment (toilaluan) [inline services/rewarding/open_category_reward.py:247]: should add clipping range (0,1) in case score < min_val

File: services/rewarding/open_category_reward.py [hunk 6]
```diff
@@ -0,0 +1,275 @@
...
+
+        return scores, sorted_questions
+
+class IQA:
+    def __init__(self, model_name="nima-vgg16-ava"):
+        import pyiqa
+        device = torch.device("cuda") if torch.cuda.is_available() else torch.device("cpu")
+        self.model = pyiqa.create_metric(model_name, device=device)
+    def __call__(self, image_path):
+        return self.model(image_path)
```
Comment (toilaluan) [inline services/rewarding/open_category_reward.py:234]: i'm thinking of miner should return PNG formatted image, currently is JPG for fixed category. And we should verify it by pillow function
https://stackoverflow.com/questions/32679589/how-to-get-the-format-of-image-with-pil

About the resolution, we should assert miner return size in one of:
```python
{
    '1:1': [
        1024,
        1024
    ],
    '3:2': [
        1152,
        768
    ],
    '2:3': [
        768,
        1152
    ],
    '19:13': [
        1216,
        832
    ],
    '13:19': [
        832,
        1216
    ],
    '16:9': [
        1360,
        768
    ],
    '9:16': [
        768,
        1360
    ]
}
```

**Comment by toilaluan** [comment]
Overall it looks very great. We should add `tests/benchmark_open_category.py` that allow miner to benchmark their endpoint and receive scores.
N_RUN=5
- Get prompts from our server
- Call miner endpoint to get images
- Get scores from reward server
We should also add rate limit on the reward server for non validator ips

**Comment by toilaluan** [comment]
this pr mean we only add general prompt style. Can we add specific style for later updates @ViktorThink ?
